Map named with keyword causes compiler to crash

Looks like we aren’t using NameAllocator where we should.

```
Caused by: java.lang.IllegalArgumentException: not a valid name: protected
        at com.squareup.javapoet.Util.checkArgument(Util.java:64)
        at com.squareup.javapoet.FieldSpec.builder(FieldSpec.java:91)
        at com.squareup.wire.java.JavaGenerator.messageAdapter(JavaGenerator.java:877)
        at com.squareup.wire.java.JavaGenerator.generateMessage(JavaGenerator.java:622)
        at com.squareup.wire.java.JavaGenerator.generateType(JavaGenerator.java:389)
        at com.squareup.wire.JavaFileWriter.call(JavaFileWriter.java:55)
        at com.squareup.wire.JavaFileWriter.call(JavaFileWriter.java:30)
```

--- a/wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java
+++ b/wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java
@@ -867,7 +867,8 @@ public final class JavaGenerator {
     for (Field field : type.getFieldsAndOneOfFields()) {
       if (field.getType().isMap()) {
         TypeName adapterType = adapterOf(fieldType(field));
-        adapter.addField(FieldSpec.builder(adapterType, field.getName(), PRIVATE, FINAL)
+        String fieldName = nameAllocator.get(field);
+        adapter.addField(FieldSpec.builder(adapterType, fieldName, PRIVATE, FINAL)
             .initializer("$T.newMapAdapter($L, $L)", ADAPTER,
                 singleAdapterFor(field.getType().getKeyType()),
                 singleAdapterFor(field.getType().getValueType()))

--- a/wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java
+++ b/wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java
@@ -282,9 +282,9 @@ public final class JavaGenerator {
         : javaName.peerClass("Abstract" + javaName.simpleName() + "Adapter");
   }
 
-  private CodeBlock singleAdapterFor(Field field) {
+  private CodeBlock singleAdapterFor(Field field, NameAllocator nameAllocator) {
     return field.getType().isMap()
-        ? CodeBlock.of("$N", field.getName())
+        ? CodeBlock.of("$N", nameAllocator.get(field))
         : singleAdapterFor(field.getType());
   }
 
@@ -305,8 +305,8 @@ public final class JavaGenerator {
     return result.build();
   }
 
-  private CodeBlock adapterFor(Field field) {
-    CodeBlock.Builder result = singleAdapterFor(field).toBuilder();
+  private CodeBlock adapterFor(Field field, NameAllocator nameAllocator) {
+    CodeBlock.Builder result = singleAdapterFor(field, nameAllocator).toBuilder();
     if (field.isPacked()) {
       result.add(".asPacked()");
     } else if (field.isRepeated()) {
@@ -892,7 +892,7 @@ public final class JavaGenerator {
     for (Field field : type.getFieldsAndOneOfFields()) {
       int fieldTag = field.getTag();
       String fieldName = nameAllocator.get(field);
-      CodeBlock adapter = adapterFor(field);
+      CodeBlock adapter = adapterFor(field, nameAllocator);
       result.addCode("$L $L.encodedSizeWithTag($L, ", leading, adapter, fieldTag)
           .addCode((useBuilder ? "value.$L" : "$L(value)"), fieldName)
           .addCode(")");
@@ -917,7 +917,7 @@ public final class JavaGenerator {
 
     for (Field field : type.getFieldsAndOneOfFields()) {
       int fieldTag = field.getTag();
-      CodeBlock adapter = adapterFor(field);
+      CodeBlock adapter = adapterFor(field, nameAllocator);
       result.addCode("$L.encodeWithTag(writer, $L, ", adapter, fieldTag)
           .addCode((useBuilder ? "value.$L" : "$L(value)"), nameAllocator.get(field))
           .addCode(");\n");
@@ -1013,7 +1013,7 @@ public final class JavaGenerator {
 
   private CodeBlock decodeAndAssign(Field field, NameAllocator nameAllocator, boolean useBuilder) {
     String fieldName = nameAllocator.get(field);
-    CodeBlock decode = CodeBlock.of("$L.decode(reader)", singleAdapterFor(field));
+    CodeBlock decode = CodeBlock.of("$L.decode(reader)", singleAdapterFor(field, nameAllocator));
     if (field.isRepeated()) {
       return useBuilder
           ? CodeBlock.of("builder.$L.add($L)", fieldName, decode)
@@ -1075,7 +1075,7 @@ public final class JavaGenerator {
         }
       } else if (!field.getType().isScalar() && !isEnum(field.getType())) {
         if (field.isRepeated()) {
-          CodeBlock adapter = singleAdapterFor(field);
+          CodeBlock adapter = singleAdapterFor(field, nameAllocator);
           result.addStatement("$T.redactElements(builder.$N, $L)", Internal.class, fieldName,
               adapter);
         } else if (field.getType().isMap()) {
@@ -1087,7 +1087,7 @@ public final class JavaGenerator {
                 adapter);
           }
         } else {
-          CodeBlock adapter = adapterFor(field);
+          CodeBlock adapter = adapterFor(field, nameAllocator);
           if (!field.isRequired()) {
             result.addCode("if (builder.$N != null) ", fieldName);
           }

--- a/wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt
+++ b/wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt
@@ -1222,7 +1222,7 @@ class KotlinGenerator private constructor(
   // TODO add support for custom adapters.
   private fun Field.getAdapterName(nameDelimiter: Char = '.'): CodeBlock {
     return if (type!!.isMap) {
-      CodeBlock.of("%NAdapter", name)
+      CodeBlock.of("%N", name + "Adapter")
     } else {
       type!!.getAdapterName(nameDelimiter)
     }

--- a/wire-library/wire-moshi-adapter/src/test/java/com/squareup/wire/MoshiTest.java
+++ b/wire-library/wire-moshi-adapter/src/test/java/com/squareup/wire/MoshiTest.java
@@ -303,7 +303,7 @@ public class MoshiTest {
 
     KeywordKotlin keyword = new KeywordKotlin.Builder().object_("object").when_(1).build();
     String json = adapter.toJson(keyword);
-    JsonUtils.assertJsonEquals(json, "{\"object\":\"object\",\"when\":1}");
+    JsonUtils.assertJsonEquals(json, "{\"object\":\"object\",\"when\":1, \"fun\":{}}");
     KeywordKotlin parseKeyword = adapter.fromJson(json);
     assertThat(parseKeyword).isEqualTo(keyword);
   }
@@ -313,7 +313,7 @@ public class MoshiTest {
 
     KeywordJava keyword = new KeywordJava.Builder().public_(true).final_("final").build();
     String json = adapter.toJson(keyword);
-    JsonUtils.assertJsonEquals(json, "{\"final\":\"final\", \"public\":true}");
+    JsonUtils.assertJsonEquals(json, "{\"final\":\"final\", \"public\":true, \"package\":{}}");
     KeywordJava parseKeyword = adapter.fromJson(json);
     assertThat(parseKeyword).isEqualTo(keyword);
   }

--- a/wire-library/wire-moshi-adapter/src/test/java/squareup/keywords/KeywordJava.java
+++ b/wire-library/wire-moshi-adapter/src/test/java/squareup/keywords/KeywordJava.java
@@ -148,7 +148,7 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
     public int encodedSize(KeywordJava value) {
       return ProtoAdapter.STRING.encodedSizeWithTag(1, value.final_)
           + ProtoAdapter.BOOL.encodedSizeWithTag(2, value.public_)
-          + package.encodedSizeWithTag(3, value.package_)
+          + package_.encodedSizeWithTag(3, value.package_)
           + value.unknownFields().size();
     }
 
@@ -156,7 +156,7 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
     public void encode(ProtoWriter writer, KeywordJava value) throws IOException {
       ProtoAdapter.STRING.encodeWithTag(writer, 1, value.final_);
       ProtoAdapter.BOOL.encodeWithTag(writer, 2, value.public_);
-      package.encodeWithTag(writer, 3, value.package_);
+      package_.encodeWithTag(writer, 3, value.package_);
       writer.writeBytes(value.unknownFields());
     }
 
@@ -168,7 +168,7 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
         switch (tag) {
           case 1: builder.final_(ProtoAdapter.STRING.decode(reader)); break;
           case 2: builder.public_(ProtoAdapter.BOOL.decode(reader)); break;
-          case 3: builder.package_.putAll(package.decode(reader)); break;
+          case 3: builder.package_.putAll(package_.decode(reader)); break;
           default: {
             reader.readUnknownField(tag);
           }

--- a/wire-library/wire-moshi-adapter/src/test/java/squareup/keywords/KeywordJava.java
+++ b/wire-library/wire-moshi-adapter/src/test/java/squareup/keywords/KeywordJava.java
@@ -15,6 +15,7 @@ import java.lang.Object;
 import java.lang.Override;
 import java.lang.String;
 import java.lang.StringBuilder;
+import java.util.Map;
 import okio.ByteString;
 
 public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder> {
@@ -40,14 +41,24 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
   )
   public final Boolean public_;
 
-  public KeywordJava(String final_, Boolean public_) {
-    this(final_, public_, ByteString.EMPTY);
+  @WireField(
+      tag = 3,
+      keyAdapter = "com.squareup.wire.ProtoAdapter#STRING",
+      adapter = "com.squareup.wire.ProtoAdapter#STRING",
+      declaredName = "package"
+  )
+  public final Map<String, String> package_;
+
+  public KeywordJava(String final_, Boolean public_, Map<String, String> package_) {
+    this(final_, public_, package_, ByteString.EMPTY);
   }
 
-  public KeywordJava(String final_, Boolean public_, ByteString unknownFields) {
+  public KeywordJava(String final_, Boolean public_, Map<String, String> package_,
+      ByteString unknownFields) {
     super(ADAPTER, unknownFields);
     this.final_ = final_;
     this.public_ = public_;
+    this.package_ = Internal.immutableCopyOf("package_", package_);
   }
 
   @Override
@@ -55,6 +66,7 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
     Builder builder = new Builder();
     builder.final_ = final_;
     builder.public_ = public_;
+    builder.package_ = Internal.copyOf(package_);
     builder.addUnknownFields(unknownFields());
     return builder;
   }
@@ -66,7 +78,8 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
     KeywordJava o = (KeywordJava) other;
     return unknownFields().equals(o.unknownFields())
         && Internal.equals(final_, o.final_)
-        && Internal.equals(public_, o.public_);
+        && Internal.equals(public_, o.public_)
+        && package_.equals(o.package_);
   }
 
   @Override
@@ -76,6 +89,7 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
       result = unknownFields().hashCode();
       result = result * 37 + (final_ != null ? final_.hashCode() : 0);
       result = result * 37 + (public_ != null ? public_.hashCode() : 0);
+      result = result * 37 + package_.hashCode();
       super.hashCode = result;
     }
     return result;
@@ -86,6 +100,7 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
     StringBuilder builder = new StringBuilder();
     if (final_ != null) builder.append(", final=").append(Internal.sanitize(final_));
     if (public_ != null) builder.append(", public=").append(public_);
+    if (!package_.isEmpty()) builder.append(", package=").append(package_);
     return builder.replace(0, 2, "KeywordJava{").append('}').toString();
   }
 
@@ -94,7 +109,10 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
 
     public Boolean public_;
 
+    public Map<String, String> package_;
+
     public Builder() {
+      package_ = Internal.newMutableMap();
     }
 
     public Builder final_(String final_) {
@@ -107,13 +125,21 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
       return this;
     }
 
+    public Builder package_(Map<String, String> package_) {
+      Internal.checkElementsNotNull(package_);
+      this.package_ = package_;
+      return this;
+    }
+
     @Override
     public KeywordJava build() {
-      return new KeywordJava(final_, public_, super.buildUnknownFields());
+      return new KeywordJava(final_, public_, package_, super.buildUnknownFields());
     }
   }
 
   private static final class ProtoAdapter_KeywordJava extends ProtoAdapter<KeywordJava> {
+    private final ProtoAdapter<Map<String, String>> package_ = ProtoAdapter.newMapAdapter(ProtoAdapter.STRING, ProtoAdapter.STRING);
+
     public ProtoAdapter_KeywordJava() {
       super(FieldEncoding.LENGTH_DELIMITED, KeywordJava.class, "type.googleapis.com/squareup.keywords.KeywordJava");
     }
@@ -122,6 +148,7 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
     public int encodedSize(KeywordJava value) {
       return ProtoAdapter.STRING.encodedSizeWithTag(1, value.final_)
           + ProtoAdapter.BOOL.encodedSizeWithTag(2, value.public_)
+          + package.encodedSizeWithTag(3, value.package_)
           + value.unknownFields().size();
     }
 
@@ -129,6 +156,7 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
     public void encode(ProtoWriter writer, KeywordJava value) throws IOException {
       ProtoAdapter.STRING.encodeWithTag(writer, 1, value.final_);
       ProtoAdapter.BOOL.encodeWithTag(writer, 2, value.public_);
+      package.encodeWithTag(writer, 3, value.package_);
       writer.writeBytes(value.unknownFields());
     }
 
@@ -140,6 +168,7 @@ public final class KeywordJava extends Message<KeywordJava, KeywordJava.Builder>
         switch (tag) {
           case 1: builder.final_(ProtoAdapter.STRING.decode(reader)); break;
           case 2: builder.public_(ProtoAdapter.BOOL.decode(reader)); break;
+          case 3: builder.package_.putAll(package.decode(reader)); break;
           default: {
             reader.readUnknownField(tag);
           }

--- a/wire-library/wire-moshi-adapter/src/test/java/squareup/keywords/KeywordKotlin.kt
+++ b/wire-library/wire-moshi-adapter/src/test/java/squareup/keywords/KeywordKotlin.kt
@@ -134,13 +134,13 @@ class KeywordKotlin(
       override fun encodedSize(value: KeywordKotlin): Int = 
         ProtoAdapter.STRING.encodedSizeWithTag(1, value.object_) +
         ProtoAdapter.INT32.encodedSizeWithTag(2, value.when_) +
-        `fun`Adapter.encodedSizeWithTag(3, value.fun_) +
+        funAdapter.encodedSizeWithTag(3, value.fun_) +
         value.unknownFields.size
 
       override fun encode(writer: ProtoWriter, value: KeywordKotlin) {
         ProtoAdapter.STRING.encodeWithTag(writer, 1, value.object_)
         ProtoAdapter.INT32.encodeWithTag(writer, 2, value.when_)
-        `fun`Adapter.encodeWithTag(writer, 3, value.fun_)
+        funAdapter.encodeWithTag(writer, 3, value.fun_)
         writer.writeBytes(value.unknownFields)
       }
 
@@ -152,7 +152,7 @@ class KeywordKotlin(
           when (tag) {
             1 -> object_ = ProtoAdapter.STRING.decode(reader)
             2 -> when_ = ProtoAdapter.INT32.decode(reader)
-            3 -> fun_.putAll(`fun`Adapter.decode(reader))
+            3 -> fun_.putAll(funAdapter.decode(reader))
             else -> reader.readUnknownField(tag)
           }
         }

--- a/wire-library/wire-moshi-adapter/src/test/java/squareup/keywords/KeywordKotlin.kt
+++ b/wire-library/wire-moshi-adapter/src/test/java/squareup/keywords/KeywordKotlin.kt
@@ -13,6 +13,7 @@ import kotlin.Any
 import kotlin.Boolean
 import kotlin.Int
 import kotlin.String
+import kotlin.collections.Map
 import kotlin.hashCode
 import kotlin.jvm.JvmField
 import okio.ByteString
@@ -32,12 +33,21 @@ class KeywordKotlin(
   )
   @JvmField
   val when_: Int? = null,
+  @field:WireField(
+    tag = 3,
+    keyAdapter = "com.squareup.wire.ProtoAdapter#STRING",
+    adapter = "com.squareup.wire.ProtoAdapter#STRING",
+    declaredName = "fun"
+  )
+  @JvmField
+  val fun_: Map<String, String> = emptyMap(),
   unknownFields: ByteString = ByteString.EMPTY
 ) : Message<KeywordKotlin, KeywordKotlin.Builder>(ADAPTER, unknownFields) {
   override fun newBuilder(): Builder {
     val builder = Builder()
     builder.object_ = object_
     builder.when_ = when_
+    builder.fun_ = fun_
     builder.addUnknownFields(unknownFields)
     return builder
   }
@@ -48,6 +58,7 @@ class KeywordKotlin(
     return unknownFields == other.unknownFields
         && object_ == other.object_
         && when_ == other.when_
+        && fun_ == other.fun_
   }
 
   override fun hashCode(): Int {
@@ -56,6 +67,7 @@ class KeywordKotlin(
       result = unknownFields.hashCode()
       result = result * 37 + object_.hashCode()
       result = result * 37 + when_.hashCode()
+      result = result * 37 + fun_.hashCode()
       super.hashCode = result
     }
     return result
@@ -65,14 +77,16 @@ class KeywordKotlin(
     val result = mutableListOf<String>()
     if (object_ != null) result += """object_=${sanitize(object_)}"""
     if (when_ != null) result += """when_=$when_"""
+    if (fun_.isNotEmpty()) result += """fun_=$fun_"""
     return result.joinToString(prefix = "KeywordKotlin{", separator = ", ", postfix = "}")
   }
 
   fun copy(
     object_: String? = this.object_,
     when_: Int? = this.when_,
+    fun_: Map<String, String> = this.fun_,
     unknownFields: ByteString = this.unknownFields
-  ): KeywordKotlin = KeywordKotlin(object_, when_, unknownFields)
+  ): KeywordKotlin = KeywordKotlin(object_, when_, fun_, unknownFields)
 
   class Builder : Message.Builder<KeywordKotlin, Builder>() {
     @JvmField
@@ -80,6 +94,9 @@ class KeywordKotlin(
 
     @JvmField
     var when_: Int? = null
+
+    @JvmField
+    var fun_: Map<String, String> = emptyMap()
 
     fun object_(object_: String?): Builder {
       this.object_ = object_
@@ -91,9 +108,15 @@ class KeywordKotlin(
       return this
     }
 
+    fun fun_(fun_: Map<String, String>): Builder {
+      this.fun_ = fun_
+      return this
+    }
+
     override fun build(): KeywordKotlin = KeywordKotlin(
       object_ = object_,
       when_ = when_,
+      fun_ = fun_,
       unknownFields = buildUnknownFields()
     )
   }
@@ -105,30 +128,38 @@ class KeywordKotlin(
       KeywordKotlin::class, 
       "type.googleapis.com/squareup.keywords.KeywordKotlin"
     ) {
+      private val funAdapter: ProtoAdapter<Map<String, String>> =
+          ProtoAdapter.newMapAdapter(ProtoAdapter.STRING, ProtoAdapter.STRING)
+
       override fun encodedSize(value: KeywordKotlin): Int = 
         ProtoAdapter.STRING.encodedSizeWithTag(1, value.object_) +
         ProtoAdapter.INT32.encodedSizeWithTag(2, value.when_) +
+        `fun`Adapter.encodedSizeWithTag(3, value.fun_) +
         value.unknownFields.size
 
       override fun encode(writer: ProtoWriter, value: KeywordKotlin) {
         ProtoAdapter.STRING.encodeWithTag(writer, 1, value.object_)
         ProtoAdapter.INT32.encodeWithTag(writer, 2, value.when_)
+        `fun`Adapter.encodeWithTag(writer, 3, value.fun_)
         writer.writeBytes(value.unknownFields)
       }
 
       override fun decode(reader: ProtoReader): KeywordKotlin {
         var object_: String? = null
         var when_: Int? = null
+        val fun_ = mutableMapOf<String, String>()
         val unknownFields = reader.forEachTag { tag ->
           when (tag) {
             1 -> object_ = ProtoAdapter.STRING.decode(reader)
             2 -> when_ = ProtoAdapter.INT32.decode(reader)
+            3 -> fun_.putAll(`fun`Adapter.decode(reader))
             else -> reader.readUnknownField(tag)
           }
         }
         return KeywordKotlin(
           object_ = object_,
           when_ = when_,
+          fun_ = fun_,
           unknownFields = unknownFields
         )
       }

--- a/wire-library/wire-moshi-adapter/src/test/proto/keyword_java.proto
+++ b/wire-library/wire-moshi-adapter/src/test/proto/keyword_java.proto
@@ -20,4 +20,5 @@ package squareup.keywords;
 message KeywordJava {
   optional string final = 1;
   optional bool public = 2;
+  map<string, string> package = 3;
 }

--- a/wire-library/wire-moshi-adapter/src/test/proto/keyword_kotlin.proto
+++ b/wire-library/wire-moshi-adapter/src/test/proto/keyword_kotlin.proto
@@ -20,4 +20,5 @@ package squareup.keywords;
 message KeywordKotlin {
   optional string object = 1;
   optional int32 when = 2;
+  map<string, string> fun = 3;
 }

